NeoML/include/NeoML/TraditionalML/FloatVector.h

```diff
@@ -164,57 +196,65 @@ inline CFloatVector::TIterator CFloatVector::end()
 }
 
 // The dot product of two vectors
-inline double DotProduct( const CFloatVector& vector1, const CFloatVector& vector2 )
+inline double DotProduct( const CSparseFloatVectorDesc& vector1, const CSparseFloatVectorDesc& vector2 )
 {
-	NeoPresume( vector1.Size() == vector2.Size() );
 	double sum = 0;
 
-	const int size = vector1.Size();
-	const float* operand1 = vector1.GetPtr();
-	const float* operand2 = vector2.GetPtr();
-
-	for( int i = 0; i < size; i++ ) {
-		sum += static_cast<double>( operand1[i] ) * operand2[i];
+	if( vector1.Indexes == nullptr ) {
+		if( vector2.Indexes == nullptr ) {
+			const int size = min( vector1.Size, vector2.Size );
+			for( int i = 0; i < size; i++ ) {
+				sum += static_cast<double>( vector1.Values[i] ) * vector2.Values[i];
+			}
+		} else {
+			int lastPos = vector2.Size - 1;
+			while( vector2.Indexes[lastPos] >= vector1.Size ) {
```

Comment:
if all vector2.Indexes > vector1.Size you will get UB!

Reply:
refactored, fixed and tested